LanguageFeatures/Static-access-shorthand/constant_expression_A09_t01.dart

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Whether such an expression followed by more selectors is a
+/// constant expression depends on the concrete selectors and types, but can use
+/// the current rules which recursively asks about the receiver being a constant
+/// function.
+///
+/// @description Checks that a constant expression shorthand followed by `!`
+/// selector is still a constant.
+/// @author sgrekhov22@gmail.com
+/// @issue 59800
+
+// SharedOptions=--enable-experiment=enum-shorthands
+
+import '../../Utils/expect.dart';
+
+class C {
+  final String value;
+  const C(this.value);
+  const C.id(this.value);
+  const factory C.f(String value) = C;
+
+  static const C instance = const C("C instance");
```

Comment:
Could we use the type `C?` to avoid some `unnecessary_non_null_assertion`s below?

Reply:
Yes. Updated here and below.